lib/libutee/arch/arm/utee_syscalls_a32.S

```diff
@@ -52,7 +60,19 @@
         .endif
         svc #0
         pop     {r5-r7,pc}
+UNWIND( .fnend)
 	END_FUNC \name
         .endm
 
+	FUNC utee_panic, :
+UNWIND( .fnstart)
+	push	{pc}
```

Comment:
This could be merged with the instruction below.

Reply:
And I forgot `.save {pc}`! But it won't work if I do. In fact, as soon as I push pc (on its own or as part of the reg list) **and** add the proper `.save` pseudo-op(s), unwind crashes and I don't understand why. With this:
```
	FUNC utee_panic, :
UNWIND( .fnstart)
	push	{r0-r11, lr, pc}
UNWIND(	.save	{r0-r11, lr, pc})
	mov	r1, sp
	bl	__utee_panic
	/* Not reached */
UNWIND( .fnend)
	END_FUNC utee_panic
```
...I get this:
```
ERROR:   [0x0] TEE-CORE: TA panicked with code 0xbeef
DEBUG:   [0x0] TEE-CORE:get_panic_regs_a32_ta:270:                              pc=0x0x1227dc
ERROR:   [0x0] TEE-CORE: Status of TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b (0xe077c18) (active)
ERROR:   [0x0] TEE-CORE:  arch: arm  load address: 0x103000  ctx-idr: 1
ERROR:   [0x0] TEE-CORE:  stack: 0x100000 10240
ERROR:   [0x0] TEE-CORE:  region 0: va 0x100000 pa 0xe41e000 size 0x3000 flags rw-
ERROR:   [0x0] TEE-CORE:  region 1: va 0x103000 pa 0xe300000 size 0x2d000 flags r-x .ta_head .text .rodata
ERROR:   [0x0] TEE-CORE:  region 2: va 0x130000 pa 0xe32d000 size 0xb000 flags r-- .rodata .ARM.extab .ARM.extab.text.__aeabi_ldivmod .ARM.extab.text.__aeabi_uldivmod .ARM.exidx .got .dynsym .rel.got .dynamic .dynstr .hash .rel.dyn
ERROR:   [0x0] TEE-CORE:  region 3: va 0x13b000 pa 0xe338000 size 0xe6000 flags rw- .data .bss
ERROR:   [0x0] TEE-CORE:  region 4: va 0 pa 0 size 0 flags ---
ERROR:   [0x0] TEE-CORE:  region 5: va 0 pa 0 size 0 flags ---
ERROR:   [0x0] TEE-CORE:  region 6: va 0 pa 0 size 0 flags ---
ERROR:   [0x0] TEE-CORE:  region 7: va 0 pa 0 size 0 flags ---
ERROR:   [0x0] TEE-CORE: Call stack:
ERROR:   [0x0] TEE-CORE:  0x001227dc utee_panic at optee_os/lib/libutee/arch/arm/utee_syscalls_a32.S:72
ERROR:   [0x0] TEE-CORE:  0x001227dc utee_panic at optee_os/lib/libutee/arch/arm/utee_syscalls_a32.S:72
ERROR:   [0x0] TEE-CORE:  0x00000006 ???
ERROR:   [0x0] TEE-CORE: 
ERROR:   [0x0] TEE-CORE: Core data-abort at address 0x12202c8 (translation fault)
ERROR:   [0x0] TEE-CORE:  fsr 0x00000005  ttbr0 0x0e07e06a  ttbr1 0x0e07806a  cidr 0x1
ERROR:   [0x0] TEE-CORE:  cpu #0          cpsr 0x20000033
ERROR:   [0x0] TEE-CORE:  r0 0x00000008      r4 0x00138b94    r8 0x0e083c18   r12 0x0a7898f0
ERROR:   [0x0] TEE-CORE:  r1 0x00138b90      r5 0x0012276f    r9 0x0e002d80    sp 0x0e0839f0
ERROR:   [0x0] TEE-CORE:  r2 0x012202cc      r6 0x00000000   r10 0x00000000    lr 0x0e017287
ERROR:   [0x0] TEE-CORE:  r3 0x012202c8      r7 0x0e0839f8   r11 0x00000000    pc 0x0e0172c8
ERROR:   [0x0] TEE-CORE: Call stack:
ERROR:   [0x0] TEE-CORE:  0x0e0172c8 unwind_exec_insn at optee_os/core/arch/arm/kernel/unwind_arm32.c:200
ERROR:   [0x0] TEE-CORE:  0x0e017573 unwind_tab at optee_os/core/arch/arm/kernel/unwind_arm32.c:302
ERROR:   [0x0] TEE-CORE:  0x0e017645 unwind_stack_arm32 at optee_os/core/arch/arm/kernel/unwind_arm32.c:349
ERROR:   [0x0] TEE-CORE:  0x0e0176d9 print_stack_arm32 at optee_os/core/arch/arm/kernel/unwind_arm32.c:368 (discriminator 1)
ERROR:   [0x0] TEE-CORE:  0x0e0146d9 __print_stack_unwind_arm32 at optee_os/core/arch/arm/kernel/abort.c:119
ERROR:   [0x0] TEE-CORE:  0x0e014a0b __abort_print at optee_os/core/arch/arm/kernel/abort.c:357
ERROR:   [0x0] TEE-CORE:  0x0e014a51 abort_print_error at optee_os/core/arch/arm/kernel/abort.c:370
ERROR:   [0x0] TEE-CORE:  0x0e01e2bb print_panic_stack at optee_os/core/arch/arm/tee/arch_svc.c:284
ERROR:   [0x0] TEE-CORE:  0x0e01e323 tee_svc_sys_return_helper at optee_os/core/arch/arm/tee/arch_svc.c:300
ERROR:   [0x0] TEE-CORE:  0x0e002d68 tee_svc_do_call at optee_os/core/arch/arm/tee/arch_svc_a32.S:82
ERROR:   [0x0] TEE-CORE: Panic 'unhandled pageable abort' at core/arch/arm/kernel/abort.c:685 <abort_handler>
ERROR:   [0x0] TEE-CORE: Call stack:
ERROR:   [0x0] TEE-CORE:  0x0e0176e9 print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm32.c:376
ERROR:   [0x0] TEE-CORE:  0x0e026997 __do_panic at optee_os/core/kernel/panic.c:54 (discriminator 1)
ERROR:   [0x0] TEE-CORE:  0x0e014e1d abort_handler at optee_os/core/arch/arm/kernel/abort.c:686
ERROR:   [0x0] TEE-CORE:  0x0e001640 .thread_abort_generic at optee_os/core/arch/arm/kernel/thread_a32.S:775
```

Reply:
What if you do:
```
	push	{r0-r11, lr}
UNWIND(	.save	{r0-r11, lr})
        mov     lr, pc
        push    {lr}
UNWIND(	.save	{lr})
```
If that doesn't work you could go back to the original code, but use `r12` instead of `r11` to pass the value of `pc`.

Reply:
Both work. Updated with your first suggestion.